### PR TITLE
Update for deb package to sync with new installation instructions.

### PR DESCRIPTION
**Pull request by wknapik**
See: https://github.com/brave/brave-com/pull/261

Needed for https://github.com/brave/brave-browser/issues/14184.

### CODE REVIEW

File: deb/brave-keyring/DEBIAN/postinst
Comment (wknapik) [inline deb/brave-keyring/DEBIAN/postinst:8]: This isn't especially pretty, but I couldn't find a more proper way to do this. None of the `apt` tools return information about `signed-by`.

Also, is it ok to use `grep` in postinst if grep itself is not a dependency of the package? Not sure.

File: deb/brave-keyring/DEBIAN/postinst
Comment (wknapik) [inline deb/brave-keyring/DEBIAN/postinst:10]: Is there a better way to do this?

File: deb/brave-keyring/DEBIAN/control
Comment (fmarier) [inline deb/brave-keyring/DEBIAN/control:7]: Since `grep` is an essential package on Debian, there's [no need to directly depend on it](https://www.debian.org/doc/debian-policy/ch-binary.html#id4).

File: deb/brave-keyring/DEBIAN/postinst
Comment (fmarier) [reply deb/brave-keyring/DEBIAN/postinst]: For simplicity, I'd suggest only looking at `/etc/apt/sources.list.d/brave-browser*` since that's what our instructions say. Anybody not following the instructions carefully gets to do this cleanup themselves.

File: deb/make-keyring.sh
Comment (fmarier) [inline deb/make-keyring.sh:11]: We need to make sure that this file doesn't get installed on recent Debians.

If it's installed and then deleted in the `postinst` script, then it will be flagged as a missing config file by the packaging system I think.

Comment (fmarier) [reply deb/make-keyring.sh:11]: Also, one more thing to test (on Ubuntu 16.04) is whether the key ends up in the right format. We have had problems in the past where GnuPG 2 would use a newer keyring format that's not compatible with apt.

File: deb/brave-keyring/DEBIAN/control
Comment (fmarier) [inline deb/brave-keyring/DEBIAN/control:5]: We should bump this.

File: deb/brave-keyring/DEBIAN/postinst
Comment (fmarier) [inline deb/brave-keyring/DEBIAN/postinst:4]: It might be time to retire this old cleanup line. It's causing a warning about `apt-key` being deprecated I believe.

File: deb/Makefile
Comment (fmarier) [inline deb/Makefile:4]: Nit: since we're doing an explicit `mkdir -p`, this could just be `.../trusted.gpg.d` without the `/*`.

File: deb/brave-keyring/DEBIAN/postinst
Comment (fmarier) [reply deb/brave-keyring/DEBIAN/postinst:10]: I'm not aware of a better way to do it.

Comment (fmarier) [reply deb/brave-keyring/DEBIAN/postinst:10]: For simplicity, I would suggest only looking at `/etc/apt/sources.list.d/brave-browser*` since that's what our instructions use.

Users not following our instructions get to cleanup this old keyring on their own :)

**Review by fmarier — CHANGES_REQUESTED** [review]

File: deb/make-keyring.sh
Comment (wknapik) [reply deb/make-keyring.sh:11]: > We need to make sure that this file doesn't get installed on recent Debians.

I'll look into that.

> We have had problems in the past where GnuPG 2 would use a newer keyring format that's not compatible with apt.

I used gpg and gpg2 on Ubuntu 16.04 and they produced the same output.
```console
% docker run --rm -it ubuntu:16.04 sh -c 'apt-get update -qq && apt-get install -qqy gnupg2 wget >/dev/null 2>&1 && wget -q https://brave-browser-apt-release.s3.brave.com/brave-core.asc && for g in gpg gpg2; do "$g" --dearmor <brave-core.asc|sha256sum; done'
dbe94f3beeac156f3a22245a0febc34686ac696ad5bca06e3761be1c6a92c31c  -
dbe94f3beeac156f3a22245a0febc34686ac696ad5bca06e3761be1c6a92c31c  -
%
```

Comment (wknapik) [reply deb/make-keyring.sh:11]: I modified the code to use `dpkg-maintscript-helper rm_conffile` instead of `rm` (and made the old keyring a conffile, which it previously wasn't), but there's a small caveat.

When I install everything the "old way" and then install the new brave-keyring package, /etc/apt/trusted.gpg.d/brave-browser-release.gpg remains in place and /etc/apt/trusted.gpg.d/brave-browser-release.gpg.dpkg-bak is created. If I then install the new brave-keyring package again, /etc/apt/trusted.gpg.d/brave-browser-release.gpg gets deleted and /etc/apt/trusted.gpg.d/brave-browser-release.gpg.dpkg-bak remains (which is harmless and by design; it will not be used, because it doesn't have the right extension).

I have no good explanation for this behavior, but it will achieve our goal over two brave-keyring upgrades. Wdyt?

Comment (wknapik) [reply deb/make-keyring.sh:11]: I modified the code to use `dpkg-maintscript-helper rm_conffile` instead of `rm` (and made the old keyring a conffile, which it previously wasn't), but there's a small caveat.

When I install everything the "old way" and then install the new brave-keyring package, /etc/apt/trusted.gpg.d/brave-browser-release.gpg remains in place and /etc/apt/trusted.gpg.d/brave-browser-release.gpg.dpkg-bak is created. If I then install the new brave-keyring package again, /etc/apt/trusted.gpg.d/brave-browser-release.gpg gets deleted and /etc/apt/trusted.gpg.d/brave-browser-release.gpg.dpkg-bak remains (which is harmless and by design; it will not be used, because it doesn't have the right extension).

I have no good explanation for this behavior, but it will achieve our goal over two brave-keyring upgrades. Wdyt?

Comment (fmarier) [reply deb/make-keyring.sh:11]: I'm wondering whether we can find a simpler approach which doesn't involve working around dpkg.

Would the following work?

0. Remove `brave-browser-release.gpg` from `conffiles`.
1. Install `brave-browser-release.gpg` into `/usr/share/brave-keyring/` in all cases.
2. In the `postinst` script, symlink `/usr/share/brave-keyring/brave-browser-release.gpg` into `/etc/apt/trusted.gpg.d/` if on an old distro. This would go after the deletion of any existing file of that name like you already have, but using just `rm` directly.
3. In the `postrm` script, delete the `/etc/apt/trusted.gpg.d/brave-browser-release.gpg` symlink if it exists.

Comment (wknapik) [reply deb/make-keyring.sh:11]: Yeah, that all seems fine, but I'm not sure how appropriate creating files in maintainer scripts is. I was trying to avoid it, but given that the approach with conffiles doesn't work as cleanly, as I'd hope, this seems like the way to go.

Comment (wknapik) [reply deb/make-keyring.sh]: Ok, ready for review.

**Comment by wknapik** [comment]
@bkero/@fmarier I tested this, but additional testing by someone else would be welcome.

File: deb/brave-keyring/DEBIAN/postrm
Comment (fmarier) [inline deb/brave-keyring/DEBIAN/postrm:4]: Nit: the `if` is likely unnecessary. Either the file is from an old package (in which case it should be deleted) or it's the symlink we added (it should also be deleted). I can't think of a case where we'd want an old keyring to stick around when the `brave-keyring` package is removed.

**Review by fmarier — COMMENTED** [review]
Do we need both `brave-browser-release.asc` and `brave-browser-release.gpg`? Does the de-armored version also work with the `signed-by` option?

File: deb/brave-keyring/DEBIAN/postrm
Comment (wknapik) [reply deb/brave-keyring/DEBIAN/postrm]: Fixed.

**Comment by wknapik** [comment]
@fmarier 
> Do we need both brave-browser-release.asc and brave-browser-release.gpg? Does the de-armored version also work with the signed-by option?

Everything starts with the user manually adding our repo. The instructions for that involve downloading brave-browser-release.asc to /usr/share/keyrings. Just downloading it, without manual dearmoring, removes the need for the user to have gnupg installed and simplifies the instructions.

After the user adds our repo as instructed, they have a reference to /usr/share/keyrings/brave-browser-release.asc in /etc/apt/sources.list.d/brave-browser-release.list via `signed-by`, so we can't delete that file.

Removing the .gpg file is problematic and modifying /etc/apt/sources.list.d/brave-browser-release.list from a maintainer script doesn't seem right to me (nor does putting a dearmored key in a file with an .asc extension).

That's why we store both files in /usr/share/keyrings.

An alternative would be to publish also the dearmored .gpg files for downloading, like I suggested initially, and then we get to only use that - both in manual instructions and in the brave-keyring package.

**Comment by fmarier** [comment]
> An alternative would be to publish also the dearmored .gpg files for downloading, like I suggested initially, and then we get to only use that - both in manual instructions and in the brave-keyring package.

That's what I was thinking, sorry I missed that the first time. We could just have the `.gpg` binary file.

Would it work in both cases? i.e. changing the `curl` line in our installation instructions to:

- Old distros: `curl https://.../brave-browser-release.gpg > /etc/apt/trusted.gpg.d/brave-browser-release.gpg`
- New distros: `curl https://.../brave-browser-release.gpg > /usr/share/keyrings/brave-browser-release.gpg`

Of course, we'd still need to include the ASCII-armored version on https://brave.com/signing-keys/, but I think that's still a manual process anyways.

**Comment by wknapik** [comment]
@fmarier yes, we can just use the .gpg files in both the manual instructions and the brave-keyring package. The only reason I was using the ascii-armored keys, was to keep the manual instructions simple and requiring as few dependencies, as possible. But with the .gpg files published for downloading, we can just use those, with no downside.

The different approaches are not between older and newer distros, but rather between systems where `signed-by` is used and ones where it's not. If it's not used, we have to have our key in /etc/apt/trusted.gpg.d, if it is, we just need the key referred by `signed-by` to be present in /usr/share/keyrings.

So we still need the symlink hack, but with a slight modification. We package only /usr/share/keyrings/brave-browser-release.gpg and on systems that use `signed-by` we remove /etc/apt/trusted.gpg.d/brave-browser-release.gpg, while on systems that don't, we replace /etc/apt/trusted.gpg.d/brave-browser-release.gpg with a symlink to /usr/share/keyrings/brave-browser-release.gpg.

Wrt publishing the keys - we'd have to do it for all three keys/channels/repos. We can either use the same urls (just ending in .gpg, instead of .asc), or we can use this opportunity to bring us closer to the Debian guidelines:
> Repositories MUST be signed with an OpenPGP key. A binary export (gpg --export) of the key SHOULD be available at the root of the repository under the filename deriv-archive-keyring.gpg, where deriv is the a short name for the repository. The file SHOULD NOT be ascii-armored (gpg --export --armor) although a separate armored version MAY be available under deriv-archive-keyring.asc.

As for the ascii-armored keys at  https://brave.com/signing-keys/ - that's perfectly fine, they're the same keys, just in a human-readable format, so nothing needs to change.

**Comment by wknapik** [comment]
Also, weirdly, there is no brave-keyring package/dependency for beta and nightly, which would also be something to look at.

**Comment by fmarier** [comment]
> Wrt publishing the keys - we'd have to do it for all three keys/channels/repos. We can either use the same urls (just ending in .gpg, instead of .asc), or we can use this opportunity to bring us closer to the Debian guidelines:
> 
> > Repositories MUST be signed with an OpenPGP key. A binary export (gpg --export) of the key SHOULD be available at the root of the repository under the filename deriv-archive-keyring.gpg, where deriv is the a short name for the repository. The file SHOULD NOT be ascii-armored (gpg --export --armor) although a separate armored version MAY be available under deriv-archive-keyring.asc.

That's a good idea. We may as well move our key to that location (that's probably a separate brave/devops issue).

**Comment by fmarier** [comment]
So to summarize, it looks like the simplest approach would be:
- replace `https://brave-browser-apt-release.s3.brave.com/brave-core.asc` with `https://brave-browser-apt-release.s3.brave.com/brave-browser-archive-keyring.gpg` (not ASCII-armored) -- needs new brave/devops issue
- update the installation instructions to just `curl` the keyring into the appropriate directory depending on distro version, removing the use of `apt-key` (and the dependency on `gnupg`?) -- https://github.com/brave/brave-com/pull/261
- only ship the binary version of the keyring in this PR

**Comment by wknapik** [comment]
> replace https://brave-browser-apt-release.s3.brave.com/brave-core.asc with https://brave-browser-apt-release.s3.brave.com/brave-browser-archive-keyring.gpg (not ASCII-armored) -- needs new brave/devops issue

Well, I wouldn't say replace. I would just add https://brave-browser-apt-release.s3.brave.com/brave-browser-archive-keyring.gpg and the other file should probably stick around, at least for a while.

> update the installation instructions to just curl the keyring into the appropriate directory depending on distro version, removing the use of apt-key (and the dependency on gnupg?) -- brave/brave-com#261

Yep.

> only ship the binary version of the keyring in this PR

Yep.

**Comment by fmarier** [comment]
> Well, I wouldn't say replace. I would just add https://brave-browser-apt-release.s3.brave.com/brave-browser-archive-keyring.gpg and the other file should probably stick around, at least for a while.

Good point. If we have the ability to add a `301` redirect there, that would be ideal.

**Comment by wknapik** [comment]
> Good point. If we have the ability to add a 301 redirect there, that would be ideal.

It came to my mind, but the files are in different formats. And the old url, ending with .asc, indicates it's plain-text.

**Comment by fmarier** [comment]
> > Good point. If we have the ability to add a 301 redirect there, that would be ideal.
> 
> It came to my mind, but the files are in different formats. And the old url, ending with .asc, indicates it's plain-text.

The name (prior to the redirect) would be misleading, but the old instructions would continue to work:
```
$ sudo apt-key --keyring /etc/apt/trusted.gpg.d/test.gpg add /etc/apt/trusted.gpg.d/brave-browser-release.gpg
OK
$ apt-key list
...
/etc/apt/trusted.gpg.d/test.gpg
-------------------------------
pub   rsa4096 2018-10-15 [SC] [expires: 2025-03-17]
      D8BA D4DE 7EE1 7AF5 2A83  4B2D 0BB7 5829 C2D4 E821
uid           [ unknown] Brave Software <support@brave.com>
sub   rsa4096 2019-10-17 [S] [expires: 2022-05-07]
```

**Comment by wknapik** [comment]
> The name (prior to the redirect) would be misleading, but the old instructions would continue to work

The instructions would work, but it's a different thing. Any code that relies on the format would suddenly start failing.

That's a fan of strong typing an immutability talking ;]

**Comment by fmarier** [comment]
> The instructions would work, but it's a different thing. Any code that relies on the format would suddenly start failing.

For sure. Maintaining two copies of the key as you suggested is the way to go if we care about preserving use-cases that we don't know about.

My worry is that these two copies will eventually get out-of-sync (Murphy's law), especially as we won't be using the ASCII file for anything anymore.

Anyways, I'll let you decide which way you want to go with the redirect, both are fine by me. Happy to review the revised PRs once they're ready.

**Comment by wknapik** [comment]
@fmarier I only want to keep the old file for a relatively short time. Once we publish the new brave-keyring package and manual instructions (I assume those are the only places where we refer to that key), we can give it, say, a month and remove the key from the old location. We can even mention it on the page with manual instructions.

**Comment by wknapik** [comment]
I downloaded the keys via https (following the instructions at brave.com/linux):
```
f329dcb22d127c6634c7a7559d2e1f39  http/beta/brave-core-nightly.asc
f329dcb22d127c6634c7a7559d2e1f39  http/nightly/brave-core-nightly.asc
999e7863a2b205129ba807b225e5e9ef  http/release/brave-core.asc
```
And from the s3 buckets behind those urls (notice, there's also a bucket for `dev`):
```
f329dcb22d127c6634c7a7559d2e1f39  s3/beta/brave-core-nightly.asc
f329dcb22d127c6634c7a7559d2e1f39  s3/dev/brave-core-nightly.asc
f329dcb22d127c6634c7a7559d2e1f39  s3/nightly/brave-core-nightly.asc
999e7863a2b205129ba807b225e5e9ef  s3/release/brave-core.asc
```
And generated new .asc and .gpg files:
```
999e7863a2b205129ba807b225e5e9ef  brave-browser-archive-keyring.asc
48f3db92ecc3bb4443929245693e900b  brave-browser-archive-keyring.gpg
f329dcb22d127c6634c7a7559d2e1f39  brave-browser-beta-archive-keyring.asc
4b3844125c29411c38e3ddb37e24a58f  brave-browser-beta-archive-keyring.gpg
f329dcb22d127c6634c7a7559d2e1f39  brave-browser-dev-archive-keyring.asc
4b3844125c29411c38e3ddb37e24a58f  brave-browser-dev-archive-keyring.gpg
f329dcb22d127c6634c7a7559d2e1f39  brave-browser-nightly-archive-keyring.asc
4b3844125c29411c38e3ddb37e24a58f  brave-browser-nightly-archive-keyring.gpg
```

**Comment by wknapik** [comment]
@fmarier updated to use new keys.

File: deb/brave-keyring/DEBIAN/postinst
Comment (bkero) [inline deb/brave-keyring/DEBIAN/postinst:11]: If this matches and this file does not exist, the script will throw an error here (but exit 0 because `-f`). I know it works, but I can imagine a user filing a bug report about this error being thrown on a brand new installation.
```
# apt install ./brave-keyring_1.11-1.deb
Reading package lists... Done
Building dependency tree
Reading state information... Done
Note, selecting 'brave-keyring' instead of './brave-keyring_1.11-1.deb'
The following NEW packages will be installed:
  brave-keyring
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/6500 B of archives.
After this operation, 0 B of additional disk space will be used.
Get:1 /deb/output/brave-keyring_1.11-1.deb brave-keyring all 1.11 [6500 B]
debconf: delaying package configuration, since apt-utils is not installed
Selecting previously unselected package brave-keyring.
(Reading database ... 4786 files and directories currently installed.)
Preparing to unpack .../brave-keyring_1.11-1.deb ...
Unpacking brave-keyring (1.11) ...
Setting up brave-keyring (1.11) ...
grep: /etc/apt/sources.list.d/brave-browser-release.list: No such file or directory
```

File: deb/brave-keyring/DEBIAN/postinst
Comment (bkero) [inline deb/brave-keyring/DEBIAN/postinst:4]: We manage the removal of old keys here. If we plan to stop doing that we should run it by @fmarier

**Review by bkero — CHANGES_REQUESTED** [review]

File: deb/brave-keyring/DEBIAN/postinst
Comment (fmarier) [reply deb/brave-keyring/DEBIAN/postinst:4]: I was the one who suggested that. My thinking is that it's been more than a year since we updated our instructions to use the separate keyring in `/etc/apt/trusted.gpg.d/brave-browser-release.gpg`. The new (well, not so new anymore) instructions don't need this since we simply overwrite the file.

Comment (wknapik) [reply deb/brave-keyring/DEBIAN/postinst:4]: A year seems like more than enough.

File: deb/brave-keyring/DEBIAN/postinst
Comment (wknapik) [reply deb/brave-keyring/DEBIAN/postinst:12]: Added a test for the existence of that file.

File: deb/brave-keyring/DEBIAN/control
Comment (wknapik) [inline deb/brave-keyring/DEBIAN/control:7]: @bkero / @fmarier seems we don't need gnupg anymore. Not sure what procps was needed for, but it also seems unnecessary. Unless I'm missing something.

Comment (bkero) [reply deb/brave-keyring/DEBIAN/control:7]: https://github.com/brave/brave-release/commit/adc029f90bdc54e726c9b1d0f3ee585f8460aaef#diff-ebb66baa63687766d559524c51821e3ed81ac16e72a8cd580c1b4c2c1f4f9922

Comment (bkero) [reply deb/brave-keyring/DEBIAN/control:7]: tl;dr we use this for ensuring sysctl is installed for some sysctls we are setting. @fmarier still recommended?

Comment (fmarier) [reply deb/brave-keyring/DEBIAN/control:7]: Yes, it looks like it's no longer necessary since we're not setting the `sysctl` settings anymore.

Comment (bkero) [reply deb/brave-keyring/DEBIAN/control:7]: Sounds good, this can be removed then

**Review by bkero — APPROVED** [review]

File: deb/brave-keyring/DEBIAN/control
Comment (mihaiplesa) [inline deb/brave-keyring/DEBIAN/control:3]: can change to yourself here or remove altogether

Comment (mihaiplesa) [reply deb/brave-keyring/DEBIAN/control:3]: also from deb/Dockerfile

Comment (mihaiplesa) [reply deb/brave-keyring/DEBIAN/control:3]: if required `Brave DevOps <devops@brave.com>`

**Review by mihaiplesa — CHANGES_REQUESTED** [review]

**Review by mihaiplesa — APPROVED** [review]

File: deb/brave-keyring/DEBIAN/control
Comment (fmarier) [inline deb/brave-keyring/DEBIAN/control:3]: I'd suggest using the same email address as the `brave-browser` package's changelog and our GPG key (as well as the Google Play entry for the Brave for Android app):

    Brave Software <support@brave.com>

That way we're not drawing attention to the devops email address.

File: deb/Makefile
Comment (fmarier) [inline deb/Makefile:4]: After building the package on my machine and looking inside the `.deb` manually, I found that it still contained
`/etc/apt/trusted.gpg.d/brave-browser-release.gpg`. That's due to the fact that I had that file lying around locallly and that `make clean` didn't delete it.

I'd suggest keeping `brave-keyring/etc/apt/trusted.gpg.d` in the `clean` target to make sure that nobody accidentally ships it by using an old build directory or git branch.

**Review by fmarier — CHANGES_REQUESTED** [review]
This is not a blocker for this PR, but `make clean` doesn't work on my machine:
```
$ make clean
rm -rf output brave-keyring/usr brave-keyring/etc/apt
/bin/rm: cannot remove 'brave-keyring/usr/share/keyrings/brave-browser-archive-keyring.gpg': Permission denied
make: *** [Makefile:4: clean] Error 1
```
because of the permissions of the docker-generated files:
```
$ ls -lh
total 12K
drwxr-xr-x 2 francois francois 4.0K Mar 31 16:53 DEBIAN
drwxr-xr-x 3 francois francois 4.0K Mar 31 17:03 etc
drwxr-xr-x 3 root     root     4.0K Mar 31 16:55 usr
```

**Comment by fmarier** [comment]
I tested the following:

- Fresh install of release on 16.04 (Xenial) using the new instructions
- Fresh install of Nightly on 20.04 (Focal) using the new instructions
- Old install of release on 16.04, new brave-keyring package
- Old install of release on 20.04, new brave-keyring package

I also manually checked that the `.deb` includes the right files and that removing it removes everything.

Everything worked as expected.

File: deb/Makefile
Comment (wknapik) [reply deb/Makefile:4]: That's a good point. I changed the recipe.

Keep in mind though, because docker is used for building and the user inside the container is root, files end up being created with root ownership and calling `make clean` may not remove them anyway (`sudo make clean` would, but that's far from ideal).

The proper thing to do would be to run as an unprivileged user in the container, but I think that's out of scope for this issue.

**Comment by wknapik** [comment]
> This is not a blocker for this PR, but make clean doesn't work on my machine:
> because of the permissions of the docker-generated files:

Yep. The correct thing to do, would be to run as an unprivileged user in the container used for building, with uid/gid matching that of the calling user on the host system. The brute force solution for now is `sudo make clean` (I know, not ideal). We can improve that in a follow-up issue.

**Review by fmarier — APPROVED** [review]

**Pull request merged**
